`responses.add_callback` triggers DeprecationWarning beyond user's control

### Steps to Reproduce

My company uses responses in combination with pytest with deprecation warnings treated as failures. The dependabot upgrade from 0.16.0 to 0.17.0 failed.

This test (which I've scrubbed to avoid sharing irrelevant and company sensitive details) triggers a DeprecationWarning upon execution.

```python
    @responses.activate
    def test_something(self, caplog):
        # setup elided
    
        def request_callback(request):
            """Used to validate that the request correctly sets the status to deleted"""
            payload = json.loads(request.body)
            assert payload['StatusType'] == SOME_CONSTANT_TO_CHECK
            return 200, {}, json.dumps(SOME_RESPONSE_DATA)
    
       responses.add_callback(
            responses.POST, '/a/url/path', callback=request_callback
       )
```
Here is the warning:

```
E           DeprecationWarning: Argument 'match_querystring' is deprecated.
Use 'responses.matchers.query_param_matcher' or 'responses.matchers.query_string_matcher'
```

### Expected Result

This test makes no use of `match_querystring`. I would not expect to see a deprecation warning.


### Actual Result

A deprecation warning triggered.

I believe the reason is because `add_callback` sets `match_querystring` to `False` by default (see https://github.com/getsentry/responses/blob/0.17.0/responses/__init__.py#L719).

`_should_match_querystring`, the method where the deprecation warning is triggered, expects the more exact check of `if match_querystring_argument is not None:` (see https://github.com/getsentry/responses/blob/0.17.0/responses/__init__.py#L333)

My guess is that `_should_match_querystring` should be able to check the `False` value too to prevent disruption to the `add_callback` API.

Thanks for the help. Please let me know if I can do more to help or if more information is needed.

--- a/CHANGES
+++ b/CHANGES
@@ -5,6 +5,8 @@
 * Removed internal `_is_string` and `_ensure_str` functions
 * Removed internal `_quote` from `test_responses.py`
 * Removed internal `_matches` attribute of `RequestsMock` object.
+* Fix issue when Deprecation Warning was raised with default arguments
+  in `responses.add_callback`. See #464
 
 0.17.0
 ------
@@ -14,7 +16,7 @@
 * Fixed issue when `passthru_prefixes` persisted across tests.
   Now `add_passthru` is valid only within a context manager or for a single function and
   cleared on exit
-* Deprecate `match_querystring` argument in `Response`` and `CallbackResponse`.
+* Deprecate `match_querystring` argument in `Response` and `CallbackResponse`.
   Use `responses.matchers.query_param_matcher` or `responses.matchers.query_string_matcher`
 * Added support for non-UTF-8 bytes in `responses.matchers.multipart_matcher`
 * Added `responses.registries`. Now user can create custom registries to

--- a/CHANGES
+++ b/CHANGES
@@ -6,7 +6,7 @@
 * Removed internal `_quote` from `test_responses.py`
 * Removed internal `_matches` attribute of `RequestsMock` object.
 * Fix issue when Deprecation Warning was raised with default arguments
-  in `responses.add_callback`. See #464
+  in `responses.add_callback` due to `match_querystring`. See #464
 
 0.17.0
 ------

--- a/responses/__init__.py
+++ b/responses/__init__.py
@@ -62,6 +62,15 @@ _real_send = HTTPAdapter.send
 logger = logging.getLogger("responses")
 
 
+class FalseBool:
+    # used for backwards compatibility, see
+    # https://github.com/getsentry/responses/issues/464
+    def __bool__(self):
+        return False
+
+    __nonzero__ = __bool__
+
+
 def urlencoded_params_matcher(params):
     warn(
         "Function is deprecated. Use 'from responses.matchers import urlencoded_params_matcher'",
@@ -277,14 +286,15 @@ class BaseResponse(object):
             return False
 
         if match_querystring_argument is not None:
-            warn(
-                (
-                    "Argument 'match_querystring' is deprecated. "
-                    "Use 'responses.matchers.query_param_matcher' or "
-                    "'responses.matchers.query_string_matcher'"
-                ),
-                DeprecationWarning,
-            )
+            if not isinstance(match_querystring_argument, FalseBool):
+                warn(
+                    (
+                        "Argument 'match_querystring' is deprecated. "
+                        "Use 'responses.matchers.query_param_matcher' or "
+                        "'responses.matchers.query_string_matcher'"
+                    ),
+                    DeprecationWarning,
+                )
             return match_querystring_argument
 
         return bool(urlparse(self.url).query)
@@ -659,7 +669,7 @@ class RequestsMock(object):
         method,
         url,
         callback,
-        match_querystring=False,
+        match_querystring=FalseBool(),
         content_type="text/plain",
         match=(),
     ):

--- a/responses/__init__.pyi
+++ b/responses/__init__.pyi
@@ -22,6 +22,7 @@ from typing_extensions import Literal
 from unittest import mock as std_mock
 from urllib.parse import quote as quote
 from urllib3.response import HTTPHeaderDict # type: ignore # Not currently exposed in typestubs.
+
 from .matchers import urlencoded_params_matcher, json_params_matcher
 from .registries import FirstMatchRegistry
 
@@ -60,6 +61,9 @@ class CallList(Sequence[Call], Sized):
     def __getitem__(self, idx: int) -> Call: ...  # type: ignore [override]
     def add(self, request: PreparedRequest, response: _Body) -> None: ...
     def reset(self) -> None: ...
+
+class FalseBool:
+    def __bool__(self) -> bool: ...
 
 class BaseResponse:
     passthrough: bool = ...
@@ -130,7 +134,7 @@ class CallbackResponse(BaseResponse):
         callback: Callable[[Any], Any],
         stream: bool = ...,
         content_type: Optional[str] = ...,
-        match_querystring: bool = ...,
+        match_querystring: Union[bool, FalseBool] = ...,
         match: MatcherIterable = ...,
     ) -> None: ...
     def get_response(  # type: ignore [override]

--- a/responses/test_responses.py
+++ b/responses/test_responses.py
@@ -592,11 +592,11 @@ def test_callback_match_querystring_default_false():
         assert resp.status_code == status
         assert "foo" in resp.headers
 
-    from _pytest.outcomes import Failed
+    with pytest.warns(None) as record:
+        run()
 
-    with pytest.raises(Failed):
-        with pytest.deprecated_call():
-            run()
+    # check that no deprecation warning was raised
+    assert not record
 
     assert_reset()
 

--- a/responses/test_responses.py
+++ b/responses/test_responses.py
@@ -559,9 +559,46 @@ def test_callback():
     assert_reset()
 
 
-def test_callback_deprecated_argument():
+def test_callback_deprecated_stream_argument():
     with pytest.deprecated_call():
         CallbackResponse(responses.GET, "url", lambda x: x, stream=False)
+
+
+def test_callback_deprecated_match_querystring_argument():
+    with pytest.deprecated_call():
+        CallbackResponse(responses.GET, "url", lambda x: x, match_querystring=False)
+
+
+def test_callback_match_querystring_default_false():
+    """
+    Test to ensure that by default 'match_querystring' in 'add_callback' is set to False
+    and does not raise deprecation
+    see: https://github.com/getsentry/responses/issues/464 and related PR
+    """
+    body = b"test callback"
+    status = 200
+    params = {"hello": "world", "I am": "a big test"}
+    headers = {"foo": "bar"}
+    url = "http://example.com/"
+
+    def request_callback(_request):
+        return status, headers, body
+
+    @responses.activate
+    def run():
+        responses.add_callback(responses.GET, url, request_callback, content_type=None)
+        resp = requests.get(url, params=params)
+        assert resp.text == "test callback"
+        assert resp.status_code == status
+        assert "foo" in resp.headers
+
+    from _pytest.outcomes import Failed
+
+    with pytest.raises(Failed):
+        with pytest.deprecated_call():
+            run()
+
+    assert_reset()
 
 
 def test_callback_exception_result():

